File: 3047-find-the-largest-area-of-square-inside-two-rectangles/3047-find-the-largest-area-of-square-inside-two-rectangles.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = [](){
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
using ll = long long;
class Solution {
public:
    long long largestSquareArea(vector<vector<int>>& bottomLeft, vector<vector<int>>& topRight) {
        int n = bottomLeft.size();
        ll ans = 0;
        for(int i = 0; i < n; i++) {
            for(int j = i+1; j < n; j++) {
                int blX = max(bottomLeft[i][0], bottomLeft[j][0]);
                int blY = max(bottomLeft[i][1], bottomLeft[j][1]);

                int trX = min(topRight[i][0], topRight[j][0]);
                int trY = min(topRight[i][1], topRight[j][1]);

                if(blX < trX && blY < trY) {
                    ll side = min(trX - blX, trY - blY);
                    ans = max(ans, side * side);
                }
            }
        } 

        return ans;
    }
};

File: 3047-find-the-largest-area-of-square-inside-two-rectangles/3047-find-the-largest-area-of-square-inside-two-rectangles.rs
impl Solution {
    pub fn largest_square_area(bottom_left: Vec<Vec<i32>>, top_right: Vec<Vec<i32>>) -> i64 {
        let mut ans = 0i64;
        let n = bottom_left.len();

        for i in 0..n {
            for j in i+1..n {
                let bl_x = bottom_left[i][0].max(bottom_left[j][0]);
                let bl_y = bottom_left[i][1].max(bottom_left[j][1]);
                let tr_x = top_right[i][0].min(top_right[j][0]);
                let tr_y = top_right[i][1].min(top_right[j][1]);

                if bl_x < tr_x && bl_y < tr_y {
                    let side = (tr_x - bl_x).min(tr_y - bl_y) as i64;
                    ans = ans.max(side * side);
                }
            }
        }

        ans        
    }
}